sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,792 @@
+	// resource is missing (for instance, because it has been deleted), the resulting property map will be nil.
+	Read func(ctx *ProviderContext, urn resource.URN, id resource.ID,
+		props resource.PropertyMap) (resource.ID, resource.PropertyMap, error)
+
+	// Update updates an existing resource with new values.
+	Update func(ctx *ProviderContext, urn resource.URN, id resource.ID,
+		olds resource.PropertyMap, news resource.PropertyMap, timeout float64,
+		ignoreChanges []string, preview bool) (resource.PropertyMap, error)
+
+	// Delete tears down an existing resource.
+	Delete func(ctx *ProviderContext, urn resource.URN, id resource.ID, props resource.PropertyMap,
+		timeout float64) error
+
+	// Construct creates a new component resource.
+	Construct func(ctx *Context, typ, name string, inputs *ConstructInputs,
+		options ResourceOption) (ConstructResult, error)
+
+	// Invoke dynamically executes a built-in function in the provider.
+	Invoke func(ctx *ProviderContext, tok string,
+		args resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error)
+
+	// SignalCancellation asks all resource providers to gracefully shut down and abort any ongoing
+	// operations. Operation aborted in this way will return an error (e.g., `Update` and `Create`
+	// will either a creation error or an initialization error. SignalCancellation is advisory and
+	// non-blocking; it is up to the host to decide how long to wait after SignalCancellation is
+	// called before (e.g.) hard-closing any gRPC connection.
+	SignalCancellation func() error
+}
+
+// ProviderContext provides a context for logging.
+type ProviderContext struct {
+	engineConn *grpc.ClientConn
+
+	// Log provides the logging interface for the Pulumi log stream
+	Log Log
+}
+
+// newProviderContext creates a new context with connection to the engine for logging.
+func newProviderContext(ctx context.Context, engineConn *grpc.ClientConn) (*ProviderContext, error) {
+	engine := pulumirpc.NewEngineClient(engineConn)
+	log := &logState{
+		engine: engine,
+		ctx:    ctx,
+	}
+	return &ProviderContext{
+		engineConn: engineConn,
+		Log:        log,
+	}, nil
+}
+
+// Close implements io.Closer.
+func (ctx *ProviderContext) Close() error {
+	return ctx.engineConn.Close()
+}
+
+type constructInput struct {
+	value  interface{}
+	secret bool
+	deps   []Resource
+}
+
+// ConstructInputs represents the inputs associated with a call to Construct.
+type ConstructInputs struct {
+	inputs map[string]constructInput
+}
+
+// Map returns the inputs as a Map.
+func (inputs *ConstructInputs) Map() Map {
+	result := Map{}
+	for k, v := range inputs.inputs {
+		output := newOutput(anyOutputType, v.deps...)
+		output.resolve(v.value, true /*known*/, v.secret, nil)
+		result[k] = output
+	}
+	return result
+}
+
+// SetArgs sets the inputs on the given args struct.
+func (inputs *ConstructInputs) SetArgs(args interface{}) error {
```

Comment:
The `inputs` arg to the user-provided `Construct` func is typed as `*ConstructInputs`. Originally, I was going to type this argument as `pulumi.Map` (`map[string]Input`) and pass the inputs like:

```go
	inputs := Map{}
 	for k, v := range deserializedInputs {
 		output := newOutput(anyOutputType, v.deps...)
 		output.resolve(v.value, true /*known*/, v.secret, nil)
 		inputs[k] = output
 	}
```

However, this wasn't all that useful in practice. For example, one of the components I was working on had a `tags` input that was typed as `map[string]string` (though, after deserializing it was actually a `map[string]interface{}`). Inside my provided `Construct` implementation I was having to manually convert the `AnyOutput` to `StringMapOutput` via an `ApplyT`, which wasn't a great experience (having to do that for each property).

So, instead I created a `ConstructInputs` type with two methods:

- `Map()` returns the inputs as a `pulumi.Map` with all the inputs as `AnyOutput`.
- `SetArgs(args interface{})` is a convenience helper. Pass it a pointer to an input struct, and it will use reflection to determine the appropriate output type.

An example of using SetArgs:

```go
type VpcArgs struct {
	// Requests an Amazon-provided IPv6 CIDR block with a /56 prefix length for the VPC. You cannot
	// specify the range of IP addresses, or the size of the CIDR block. Default is `false`.  If set
	// to `true`, then subnets created will default to `assignIpv6AddressOnCreation: true` as well.
	AssignGeneratedIpv6CidrBlock pulumi.BoolInput `pulumi:"assignGeneratedIpv6CidrBlock"`

	// The CIDR block for the VPC. Defaults to "10.0.0.0/16" if unspecified.
	CidrBlock pulumi.StringInput `pulumi:"cidrBlock"`

	// A boolean flag to enable/disable ClassicLink
	// for the VPC. Only valid in regions and accounts that support EC2 Classic.
	// See the [ClassicLink documentation][1] for more information. Defaults false.
	EnableClassiclink pulumi.BoolInput `pulumi:"enableClassiclink"`

	// A boolean flag to enable/disable ClassicLink DNS Support for the VPC.
	// Only valid in regions and accounts that support EC2 Classic.
	EnableClassiclinkDnsSupport pulumi.BoolInput `pulumi:"enableClassiclinkDnsSupport"`

	// A boolean flag to enable/disable DNS hostnames in the VPC. Defaults to true if unspecified.
	EnableDnsHostnames pulumi.BoolInput `pulumi:"enableDnsHostnames"`

	// A boolean flag to enable/disable DNS support in the VPC. Defaults true if unspecified.
	EnableDnsSupport pulumi.BoolInput `pulumi:"enableDnsSupport"`

	// A tenancy option for instances launched into the VPC. Defaults to "default" if unspecified.
	InstanceTenancy pulumi.StringInput `pulumi:"instanceTenancy"` // TODO enum "default" | "dedicated"?

	// A mapping of tags to assign to the resource.
	Tags pulumi.StringMapInput `pulumi:"tags"`
}

// In my Construct for this component:
args := &VpcArgs{}
err := inputs.SetArgs(args)
if err != nil {
    return pulumi.ConstructResult{}, errors.Wrap(err, "setting args")
}
```

Reply:
> Does having the SetArgs helper make sense? Should we do this in some other way?

FWIW - As a consumer of this - the `SetArgs` helper feels very useful to project args into a typed input Args bag for a component.  

In fact, it made me want something similar in reverse - so that this code could be replaced with `return pulumi.ConstructOutputsFromComponent(staticPage)`:

```go
	// Return the component resource's URN and outputs as its state.
	return pulumi.ConstructResult{
		URN: staticPage.URN(),
		State: pulumi.Map{
			"bucket":     staticPage.Bucket,
			"websiteUrl": staticPage.WebsiteUrl,
		},
	}, nil
```

Doing so would reduce the number of places I have to touch when updating the public API surface area of a component from 3 to 2 (just the schema and implementation).

> I didn't want to force component authors to have to define it twice manually like this. But curious if you think we should force component authors to define these the same way as the Go codegen does.

I believe the only place the `vpcArgs` args is needed is for enabling `VpcArgs` to be an `Input` itself (with an underlying value type of `vpcArgs`).  I don't think there is any need to require component authors of Go-based multi-language components to implement that.

Would love @pgavlin or @EvanBoyle or @viveklak  confirmation on this from a Go component model perspective.

Reply:
> In fact, it made me want something similar in reverse - so that this code could be replaced with `return pulumi.ConstructOutputsFromComponent(staticPage)`:

Yes! I've wanted the exact same helper. I added `SetArgs` out of necessity, but would also love something along the lines of your suggestion. I'll add it to this PR sometime before next week when I have a spare moment.

Aside: Ultimately, we're going to want to significantly improve the experience here such that someone doesn't even have to write `Construct` at all for components. They'll just write their Go component as normal, and perhaps call some registration function, passing it the type of the args and resource, and then a schema will be able to be automatically generated from that as well as the construct implementation, so the user doesn't have to deal with any of this, and doesn't have to deal with keeping a separate schema.json up-to-date.

Reply:
> I believe the only place the `vpcArgs` args is needed is for enabling `VpcArgs` to be an `Input` itself (with an underlying value type of `vpcArgs`). I don't think there is any need to require component authors of Go-based multi-language components to implement that.

Ah, that makes sense.